File: app/src/main/java/com/settlex/android/ui/auth/login/LoginActivity.java
package com.settlex.android.ui.auth.login;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.Rect;
import android.graphics.Typeface;
import android.os.Bundle;
import android.text.Editable;
import android.text.InputType;
import android.text.SpannableString;
import android.text.Spanned;
import android.text.TextWatcher;
import android.text.style.ForegroundColorSpan;
import android.view.MotionEvent;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;

import com.settlex.android.R;
import com.settlex.android.data.remote.profile.ProfileService;
import com.settlex.android.databinding.ActivityLoginBinding;
import com.settlex.android.ui.auth.forgot_password.ForgotPasswordActivity;
import com.settlex.android.ui.auth.register.SignUpActivity;
import com.settlex.android.ui.auth.AuthViewModel;
import com.settlex.android.ui.common.components.BiometricAuthHelper;
import com.settlex.android.util.ui.ProgressLoaderController;
import com.settlex.android.ui.dashboard.DashboardActivity;
import com.settlex.android.ui.info.help.AuthHelpActivity;
import com.settlex.android.util.network.NetworkMonitor;
import com.settlex.android.util.string.StringFormatter;
import com.settlex.android.util.ui.StatusBar;
import com.settlex.android.ui.common.util.DialogHelper;

import java.util.Objects;

import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class LoginActivity extends AppCompatActivity {
    private boolean isPasswordVisible = false;

    // dependencies
    private ProgressLoaderController progressLoader;
    private ActivityLoginBinding binding;
    private AuthViewModel authViewModel;
    private boolean isConnected = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityLoginBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        authViewModel = new ViewModelProvider(this).get(AuthViewModel.class);
        progressLoader = new ProgressLoaderController(this);

        setupUiActions();
        syncUserStateWithUI();
        observeNetworkStatus();
        observeLoginStatus();
    }

    private void setupUiActions() {
        StatusBar.setStatusBarColor(this, R.color.white);
        setupEditTextFocusHandler();
        setupAuthActionTexts();
        setupInputValidation();
        setupPasswordVisibilityToggle();
        clearFocusOnLastEditTextField();

        // Click listeners
        binding.btnBackBefore.setOnClickListener(v -> getOnBackPressedDispatcher().onBackPressed());
        binding.btnSignUp.setOnClickListener(v -> navigateTo(SignUpActivity.class));
        binding.btnHelp.setOnClickListener(v -> navigateTo(AuthHelpActivity.class));
        binding.btnForgotPassword.setOnClickListener(v -> navigateTo(ForgotPasswordActivity.class));
        binding.btnSignIn.setOnClickListener(v -> attemptLogin());
        binding.btnSwitchAccount.setOnClickListener(v -> {
            authViewModel.signOut();
            showLoggedOutLayout();
        });
        binding.btnFingerprint.setOnClickListener(v -> promptBiometricsAuth());
    }

    private void syncUserStateWithUI() {
        LoginUiModel currentUser = authViewModel.getCurrentUserLiveData().getValue();
        if (currentUser == null) {
            showLoggedOutLayout();
            return;
        }

        // user is logged in
        showLoggedInLayout(
                currentUser.getPhotoUrl(),
                currentUser.getDisplayName(),
                currentUser.getEmail());

        // check user pref
        boolean isFingerPrintEnabled = Boolean.TRUE.equals(authViewModel.getLoginBiometricsEnabled().getValue());
        binding.btnFingerprint.setVisibility(isFingerPrintEnabled ? View.VISIBLE : View.GONE);
        if (isFingerPrintEnabled) promptBiometricsAuth();
    }

    private void promptBiometricsAuth() {
        if (BiometricAuthHelper.isBiometricAvailable(this)) {
            BiometricAuthHelper biometric = new BiometricAuthHelper(
                    this,
                    this,
                    new BiometricAuthHelper.BiometricAuthCallback() {
                        @Override
                        public void onAuthenticated() {
                            navigateTo(DashboardActivity.class);
                        }

                        @Override
                        public void onError(java.lang.String message) {
                        }

                        @Override
                        public void onFailed() {
                        }
                    });
            biometric.authenticate("Confirm your identity", "Use Password");
        }
    }

    // OBSERVERS ====
    private void observeNetworkStatus() {
        NetworkMonitor.getNetworkStatus().observe(this, isConnected -> {
            if (!isConnected) showNoInternetDialog();
            this.isConnected = isConnected;
        });
    }

    private void observeLoginStatus() {
        authViewModel.getLoginLiveData().observe(this, result -> {
            if (result != null) {
                switch (result.getStatus()) {
                    case LOADING -> progressLoader.show();
                    case SUCCESS -> onLoginStatusSuccess();
                    case FAILURE -> onLoginStatusError(result.getError());
                }
            }
        });
    }

    private void onLoginStatusSuccess() {
        startActivity(new Intent(this, DashboardActivity.class));
        finishAffinity();
        progressLoader.hide();
    }

    private void onLoginStatusError(java.lang.String error) {
        binding.txtError.setText(error);
        binding.txtError.setVisibility(View.VISIBLE);

        progressLoader.hide();
    }

    private void attemptLogin() {
        if (!isConnected) {
            showNoInternetDialog();
            return;
        }

        java.lang.String email = Objects.requireNonNull(binding.editTxtEmail.getText()).toString().trim();
        java.lang.String password = Objects.requireNonNull(binding.editTxtPassword.getText()).toString().trim();

        attemptUserLoginWithEmailAndPassword(email, password);
    }

    private void attemptUserLoginWithEmailAndPassword(java.lang.String email, java.lang.String password) {
        authViewModel.loginWithEmail(email, password);
    }

    private void showLoggedInLayout(java.lang.String photoUrl, java.lang.String displayName, java.lang.String email) {
        java.lang.String userDisplayName = "Hi, " + displayName.toUpperCase();
        java.lang.String userEmail = "(" + email + ")";

        ProfileService.loadProfilePic(photoUrl, binding.userProfile);
        binding.userDisplayName.setText(userDisplayName);
        binding.userEmail.setText(StringFormatter.maskEmail(userEmail));
        binding.editTxtEmail.setText(email);

        // Show
        binding.showCurrentUserLayout.setVisibility(View.VISIBLE);
        binding.btnSwitchAccount.setVisibility(View.VISIBLE);

        // Hide
        binding.logo.setVisibility(View.GONE);
        binding.txtInputLayoutEmail.setVisibility(View.GONE);
        binding.btnSignUp.setVisibility(View.GONE);
    }

    private void showLoggedOutLayout() {
        // Hide
        binding.showCurrentUserLayout.setVisibility(View.GONE);
        binding.btnFingerprint.setVisibility(View.GONE);
        binding.btnSwitchAccount.setVisibility(View.GONE);

        // Show
        binding.txtInputLayoutEmail.setVisibility(View.VISIBLE);
        binding.btnSignUp.setVisibility(View.VISIBLE);
        binding.logo.setVisibility(View.VISIBLE);
    }

    private void showNoInternetDialog() {
        java.lang.String title = "Network Unavailable";
        java.lang.String message = "Please check your Wi-Fi or cellular data and try again";

        DialogHelper.showSimpleAlertDialog(
                this,
                title,
                message);
    }

    private void setupAuthActionTexts() {
        java.lang.String fullText = "Don't have an account yet?\nClick here to register";
        java.lang.String textToHighlight = "Click here to register";

        int startIndex = fullText.indexOf(textToHighlight);
        int endIndex = fullText.length();

        SpannableString signUpText = new SpannableString(fullText);
        signUpText.setSpan(
                new ForegroundColorSpan(Color.parseColor("#0044CC")),
                startIndex,
                endIndex,
                Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
        );
        binding.btnSignUp.setText(signUpText);

        setupSwitchAccount();
    }

    private void setupSwitchAccount() {
        java.lang.String fullText = "Not you?\nSwitch Account";
        java.lang.String textToHighlight = "Switch Account";

        int startIndex = fullText.indexOf(textToHighlight);
        int endIndex = startIndex + textToHighlight.length();

        SpannableString switchText = new SpannableString(fullText);
        switchText.setSpan(
                new ForegroundColorSpan(Color.parseColor("#0044CC")),
                startIndex,
                endIndex,
                Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
        );

        binding.btnSwitchAccount.setText(switchText);
    }

    private void navigateTo(Class<? extends Activity> activityClass) {
        startActivity(new Intent(this, activityClass));
    }

    private void setupEditTextFocusHandler() {
        int focusBgRes = R.drawable.bg_edit_txt_custom_white_focused;
        int defaultBgRes = R.drawable.bg_edit_txt_custom_gray_not_focused;

        binding.editTxtPassword.setOnFocusChangeListener((v, hasFocus) -> binding.editTxtPasswordBg.setBackgroundResource(hasFocus ? focusBgRes : defaultBgRes));
    }

    private void updateSignInButtonState() {
        java.lang.String email = Objects.requireNonNull(binding.editTxtEmail.getText()).toString().trim();
        java.lang.String password = Objects.requireNonNull(binding.editTxtPassword.getText()).toString().trim();

        binding.btnSignIn.setEnabled(!email.isEmpty() && !password.isEmpty());
    }

    private void setupInputValidation() {
        TextWatcher validationWatcher = new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void afterTextChanged(Editable s) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                java.lang.String password = binding.editTxtPassword.getText().toString().trim();
                binding.passwordToggle.setVisibility((!password.isEmpty()) ? View.VISIBLE : View.GONE);
                binding.txtError.setVisibility(View.GONE);
                updateSignInButtonState();
            }
        };

        binding.editTxtEmail.addTextChangedListener(validationWatcher);
        binding.editTxtPassword.addTextChangedListener(validationWatcher);
    }

    private void setupPasswordVisibilityToggle() {
        binding.passwordToggle.setOnClickListener(v -> {
            Typeface currentTypeface = binding.editTxtPassword.getTypeface();
            isPasswordVisible = !isPasswordVisible;

            int inputType = isPasswordVisible ? InputType.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD : InputType.TYPE_TEXT_VARIATION_PASSWORD;

            binding.editTxtPassword.setInputType(InputType.TYPE_CLASS_TEXT | inputType);
            binding.passwordToggle.setImageResource(isPasswordVisible ? R.drawable.ic_visibility_on_filled : R.drawable.ic_visibility_off_filled);

            binding.editTxtPassword.setTypeface(currentTypeface);
            binding.editTxtPassword.setSelection(binding.editTxtPassword.getText().length());
        });
    }

    private void clearFocusOnLastEditTextField() {
        binding.editTxtPassword.setOnEditorActionListener((v, actionId, event) -> {
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                // Hide the keyboard
                InputMethodManager imm = (InputMethodManager) v.getContext().getSystemService(Context.INPUT_METHOD_SERVICE);
                imm.hideSoftInputFromWindow(v.getWindowToken(), 0);

                // Clear focus
                v.clearFocus();
                return true;
            }
            return false;
        });
    }


    @Override
    public boolean dispatchTouchEvent(MotionEvent event) {
        if (event.getAction() == MotionEvent.ACTION_DOWN) {
            View v = getCurrentFocus();
            if (v instanceof EditText) {
                Rect outRect = new Rect();
                v.getGlobalVisibleRect(outRect);
                if (!outRect.contains((int) event.getRawX(), (int) event.getRawY())) {
                    v.clearFocus();
                    hideKeyboard(v);
                }
            }
        }
        return super.dispatchTouchEvent(event);
    }

    private void hideKeyboard(View view) {
        InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
        if (imm != null) {
            imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
        }
    }
}

File: app/src/main/java/com/settlex/android/ui/auth/login/LoginActivity.kt
package com.settlex.android.ui.auth.login

import android.app.Activity
import android.content.Intent
import android.graphics.Rect
import android.os.Bundle
import android.text.Editable
import android.text.InputType
import android.text.SpannableString
import android.text.Spanned
import android.text.TextWatcher
import android.text.style.ForegroundColorSpan
import android.view.MotionEvent
import android.view.View
import android.view.View.OnFocusChangeListener
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import android.widget.EditText
import android.widget.TextView
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.core.graphics.toColorInt
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.settlex.android.R
import com.settlex.android.data.remote.profile.ProfileService
import com.settlex.android.databinding.ActivityLoginBinding
import com.settlex.android.ui.auth.AuthViewModel
import com.settlex.android.ui.auth.forgot_password.ForgotPasswordActivity
import com.settlex.android.ui.auth.register.RegisterActivity
import com.settlex.android.ui.common.components.BiometricAuthHelper
import com.settlex.android.ui.common.components.BiometricAuthHelper.BiometricAuthCallback
import com.settlex.android.ui.common.state.UiState
import com.settlex.android.ui.dashboard.DashboardActivity
import com.settlex.android.util.string.StringFormatter
import com.settlex.android.util.ui.ProgressLoaderController
import com.settlex.android.util.ui.StatusBar
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch

@AndroidEntryPoint
class LoginActivity : AppCompatActivity() {
    private var isPasswordVisible = false

    // Dependencies
    private val progressLoader: ProgressLoaderController by lazy { ProgressLoaderController(this) }
    private lateinit var binding: ActivityLoginBinding
    private val authViewModel: AuthViewModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityLoginBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setupUiActions()
        observeLoginEvent()
        syncUserStateWithUI()
        setupClickListeners()
    }

    private fun setupUiActions() {
        StatusBar.setColor(this, R.color.white)
        setupAuthActionTexts()
        setupInputValidation()
        setupEditTextFocusHandler()
        setupPasswordInputDoneAction()
    }

    private fun setupClickListeners() {
        binding.btnSignUp.setOnClickListener {
            navigateTo(
                RegisterActivity::class.java
            )
        }
        binding.btnForgotPassword.setOnClickListener {
            navigateTo(
                ForgotPasswordActivity::class.java
            )
        }
        binding.btnBackBefore.setOnClickListener { onBackPressedDispatcher.onBackPressed() }
        binding.btnSwitchAccount.setOnClickListener { showUnauthenticatedUi() }
        binding.btnFingerprint.setOnClickListener { promptBiometricsAuth() }
        binding.btnTogglePassword.setOnClickListener { togglePasswordVisibility() }
        binding.btnSignIn.setOnClickListener { attemptLogin() }
    }

    private fun syncUserStateWithUI() {
        val userState = authViewModel.userState.value
        when (userState) {
            is LoginState.NoUser -> {
                showUnauthenticatedUi()
            }

            is LoginState.CurrentUser -> {
                showAuthenticatedUi(
                    userState.photoUrl,
                    userState.displayName,
                    userState.email
                )
            }
        }

        // check user pref
        // val isFingerPrintEnabled =
        // Boolean.TRUE == authViewModel.getLoginBiometricsEnabled().getValue()
        // binding.btnFingerprint.visibility = if (isFingerPrintEnabled) View.VISIBLE else View.GONE
        // if (isFingerPrintEnabled) promptBiometricsAuth()
    }

    private fun showAuthenticatedUi(photoUrl: String?, displayName: String, email: String) {
        val formattedDisplayName = "Hi, ${displayName.uppercase()}"
        val formattedEmail = "( $email )"

        ProfileService.loadProfilePic(photoUrl, binding.userProfile)
        binding.tvUserDisplayName.text = formattedDisplayName
        binding.tvUserEmail.text = StringFormatter.maskEmail(email)
        binding.etEmail.setText(formattedEmail)

        // Show
        binding.showCurrentUserLayout.visibility = View.VISIBLE
        binding.btnSwitchAccount.visibility = View.VISIBLE

        // Hide
        binding.logo.visibility = View.GONE
        binding.tilEmail.visibility = View.GONE
        binding.btnSignUp.visibility = View.GONE
    }

    private fun showUnauthenticatedUi() {
        // Hide
        binding.showCurrentUserLayout.visibility = View.GONE
        binding.btnFingerprint.visibility = View.GONE
        binding.btnSwitchAccount.visibility = View.GONE

        // Show
        binding.tilEmail.visibility = View.VISIBLE
        binding.btnSignUp.visibility = View.VISIBLE
        binding.logo.visibility = View.VISIBLE
    }

    // Observers
    private fun observeLoginEvent() {
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                authViewModel.loginEvent.collect {
                    when (it) {
                        is UiState.Loading -> progressLoader.show()
                        is UiState.Success -> onLoginSuccess()
                        is UiState.Failure -> onLoginFailure(it.exception.message)
                    }
                }
            }
        }
    }

    private fun onLoginSuccess() {
        startActivity(Intent(this, DashboardActivity::class.java))
        finishAffinity()

        progressLoader.hide()
    }

    private fun onLoginFailure(error: String?) {
        binding.tvError.text = error
        binding.tvError.visibility = View.VISIBLE

        progressLoader.hide()
    }

    private fun attemptLogin() {
        val email = binding.etEmail.text.toString().trim()
        val password = binding.etPassword.text.toString().trim()

        authViewModel.login(email, password)
    }

    private fun promptBiometricsAuth() {
        if (BiometricAuthHelper.isBiometricAvailable(this)) {
            val biometric = BiometricAuthHelper(
                this,
                this,
                object : BiometricAuthCallback {
                    override fun onAuthenticated() {
                        navigateTo(DashboardActivity::class.java)
                    }

                    override fun onError(message: String?) {
                    }

                    override fun onFailed() {
                    }
                })
            biometric.authenticate("Confirm your identity", "Use Password")
        }
    }

    private fun togglePasswordVisibility() {
        val currentTypeface = binding.etPassword.typeface
        isPasswordVisible = !isPasswordVisible

        val inputType = if (isPasswordVisible) {
            InputType.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD
        } else {
            InputType.TYPE_TEXT_VARIATION_PASSWORD
        }

        binding.etPassword.inputType = InputType.TYPE_CLASS_TEXT or inputType
        binding.btnTogglePassword.setImageResource(if (isPasswordVisible) R.drawable.ic_visibility_on_filled else R.drawable.ic_visibility_off_filled)

        binding.etPassword.setTypeface(currentTypeface)
        binding.etPassword.setSelection(binding.etPassword.getText()!!.length)
    }

    private fun setupAuthActionTexts() {
        val fullText = "Don't have an account yet?\nClick here to register"
        val textToHighlight = "Click here to register"

        val startIndex = fullText.indexOf(textToHighlight)
        val endIndex = fullText.length

        val signUpText = SpannableString(fullText)
        signUpText.setSpan(
            ForegroundColorSpan("#0044CC".toColorInt()),
            startIndex,
            endIndex,
            Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
        )
        binding.btnSignUp.text = signUpText

        setupSwitchAccount()
    }

    private fun setupSwitchAccount() {
        val fullText = "Not you?\nSwitch Account"
        val textToHighlight = "Switch Account"

        val startIndex = fullText.indexOf(textToHighlight)
        val endIndex = startIndex + textToHighlight.length

        val switchText = SpannableString(fullText)
        switchText.setSpan(
            ForegroundColorSpan("#0044CC".toColorInt()),
            startIndex,
            endIndex,
            Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
        )

        binding.btnSwitchAccount.text = switchText
    }

    private fun navigateTo(activityClass: Class<out Activity>) {
        startActivity(Intent(this, activityClass))
    }

    private fun setupEditTextFocusHandler() {
        val focusBgRes = R.drawable.bg_edit_txt_custom_white_focused
        val defaultBgRes = R.drawable.bg_edit_txt_custom_gray_not_focused

        binding.etPassword.onFocusChangeListener =
            OnFocusChangeListener { _, hasFocus: Boolean ->
                binding.etPasswordBg.setBackgroundResource(if (hasFocus) focusBgRes else defaultBgRes)
            }
    }

    private fun updateSignInButtonState() {
        val email = binding.etEmail.text.toString().trim()
        val password = binding.etPassword.text.toString().trim()

        binding.btnSignIn.isEnabled = email.isNotEmpty() && password.isNotEmpty()
    }

    private fun setupInputValidation() {
        val validationWatcher: TextWatcher = object : TextWatcher {
            override fun beforeTextChanged(
                s: CharSequence?,
                start: Int,
                count: Int,
                after: Int
            ) {
            }

            override fun afterTextChanged(s: Editable?) {
            }

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                val password = binding.etPassword.getText().toString().trim()
                binding.btnTogglePassword.visibility =
                    if (password.isNotEmpty()) View.VISIBLE else View.GONE
                binding.tvError.visibility = View.GONE
                updateSignInButtonState()
            }
        }

        binding.etEmail.addTextChangedListener(validationWatcher)
        binding.etPassword.addTextChangedListener(validationWatcher)
    }

    private fun setupPasswordInputDoneAction() {
        binding.etPassword.setOnEditorActionListener { v: TextView?, actionId: Int, _ ->
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                hideKeyboard(v!!)
                v.clearFocus()
                return@setOnEditorActionListener true
            }
            false
        }
    }

    override fun dispatchTouchEvent(event: MotionEvent): Boolean {
        if (event.action == MotionEvent.ACTION_DOWN) {
            val v = currentFocus
            if (v is EditText) {
                val outRect = Rect()
                v.getGlobalVisibleRect(outRect)
                if (!outRect.contains(event.rawX.toInt(), event.rawY.toInt())) {
                    v.clearFocus()
                    hideKeyboard(v)
                }
            }
        }
        return super.dispatchTouchEvent(event)
    }

    private fun hideKeyboard(view: View) {
        val imm = getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager?
        imm!!.hideSoftInputFromWindow(view.windowToken, 0)
    }
}
